clash-cores/src/Clash/Cores/Crc/Internal.hs

```diff
@@ -0,0 +1,737 @@
+  -> SNat nLanes
+  -- ^ The number of lanes
+  -> TH.Q [TH.Dec]
+deriveHardwareCrc crc dataWidth@SNat nLanes@SNat = do
+  let crcTy = pure $ typeRepToTHType $ someTypeRep (Proxy @crc)
+      dataWidthTy = pure $ TH.LitT $ TH.NumTyLit $ snatToNum dataWidth
+      nLanesTy = pure $ TH.LitT $ TH.NumTyLit $ snatToNum nLanes
+      hwParams = lift $ mkCrcHardwareParams crc dataWidth nLanes
+
+  [d| instance HardwareCrc $crcTy $dataWidthTy $nLanesTy where
+        crcHardwareParams _ _ _ = $hwParams |]
+
+-- | Matrix multiplication in the binary finite field
+matVecMul
+  :: forall (m :: Nat)
+            (n :: Nat)
+   . KnownNat m
+  => KnownNat n
+  -- ^ @n@ columns
+  => Vec m (BitVector n)
+  -- ^ The Matrix
+  -> BitVector n
+  -- ^ The column vector
+  -> BitVector m
+  -- ^ The resulting vector
+matVecMul mat vec = v2bv $ fmap (\row -> reduceXor $ vec .&. row) mat
+
+-- | Like 'subSNat' but uses a 'Data.Type.Ord.<=' constraint
+subSNatLe
+  :: forall (n :: Nat)
+            (m :: Nat)
+   . m <= n
+  => SNat n
+  -> SNat m
+  -> SNat (n - m)
+subSNatLe a b = leToPlus @m @n $ subSNat a b
+
+-- | Like 'take' but uses a 'Data.Type.Ord.<=' constraint
+takeLe
+  :: forall (n :: Nat)
+            (m :: Nat)
+            (a :: Type)
+   . n <= m
+  => SNat n
+  -- ^ How many elements to take
+  -> Vec m a
+  -- ^ input vector
+  -> Vec n a
+takeLe SNat vs = leToPlus @n @m $ takeI vs
+
+-- | Apply a step to a CRC and input given by @lanePrev@
+--
+-- This function is meant to be used via 'smap' and crashes if called with
+-- out-of-bounds @lanePrev@.
+laneStep
+  :: CrcHardwareParams crcWidth dataWidth nLanes
+  -> SNat lanePrev
+  -> BitVector crcWidth
+  -> Vec nLanes (BitVector dataWidth)
+  -> BitVector crcWidth
+laneStep (CrcHardwareParams SNat nLanes@SNat (CrcParams SNat _ _ _ _ _) laneParams _) lanePrev@SNat crc input
+  = let lane = addSNat lanePrev d1 in case compareSNat lane nLanes of
+    SNatLE -> matVecMul (getFGMatrix laneParams lane) (crc ++# (pack $ takeLe lane input))
+    _      -> clashCompileError "laneStep: Absurd, Report this to the Clash compiler team: https://github.com/clash-lang/clash-compiler/issues"
+
+-- | A parallel multilane CRC engine and validator
+--
+-- @dataWidth@ should /always/ cleanly divide @crcWidth@. If this is not the case
+-- then it is possible for the last message to be misaligned.
+--
+-- For example if you have a @dataWidth@ of @48@ and a @crcWidth@ of @32@. Then a possible
+-- message could be 128 bits, two words + CRC. But this cannot be safely
+-- run through the CRC circuit, because the final fragment contains 32 valid bits
+-- and 16 invalid bits.
+--
+-- This rule is __NOT__ checked at the type level because as long you take
+-- care that misalignment does not occur it is not a problem.
+crcEngine
+  :: forall (dom :: Domain)
+            (crc :: Type)
+            (dataWidth :: Nat)
+            (nLanes :: Nat)
+   . HiddenClockResetEnable dom
```

Comment:
Didn't you intend to eliminate the reset input of both `crcEngine` and `crcValidator`?